machine/machine_core/ssh_connection.py

```diff
@@ -307,15 +307,13 @@ def execute(self, command=None, script=None, input=None, environment={},
         additional_ssh_params += self.__execution_opts(direct=direct)
 
         if command:
-            if getattr(command, "strip", None):  # Is this a string?
-                cmd += [command]
-                if not quiet:
-                    self.message("+", command)
-            else:
+            if isinstance(command, list):
                 # use shlex.join() once Python 3.8 is available everywhere
-                cmd.append(' '.join(shlex.quote(arg) for arg in command))
-                if not quiet:
-                    self.message("+", *command)
+                command = ' '.join(shlex.quote(arg) for arg in command)
```

Comment:
I like the `isinstance()` (although `list` is probably too specific — tuple might easily be used here).  I don't like replacing `command` with a new value of a different type.  I think this would be better with two separate `.append()` calls.

If you'd prefer to keep the DRY savings here, then please a new variable (but then you'll need an else branch again).

Reply:
hmm, I'd like to rename the argument to `command_str_or_list`, but that's an API break. I wanted to avoid having "cmd", "command", and "command_string". In that case I'll just replace the getattr with an isinstance(str) (no basestr any more in py 3 \o/).

Reply:
Hmm, no -- I really *really* want to get rid of that `else` branch, as it's just awkward to duplicate the `set -e` addition as well. I still don't understand why it's bad to transform one allowed data type variation into the canonical form..

Reply:
Done without shadowing, and assigning to `command_str` now.

Reply:
Dropped that commit.